kolibri/content/management/commands/importcontent.py

```diff
@@ -152,16 +153,21 @@ def _transfer(self, method, channel_id, path=None, node_ids=None, exclude_node_i
                     filetransfer = transfer.FileCopy(srcpath, dest)
 
                 finished = False
-                while not finished:
-                    finished, increment = self._start_file_transfer(
-                        f, filetransfer, overall_progress_update)
-                    if self.is_cancelled():
-                        self.cancel()
-                        break
-                    if increment == 2:
-                        file_checksums_to_annotate.append(f.id)
-                    else:
-                        number_of_skipped_files += increment
+                try:
+                    while not finished:
+                        finished, increment = self._start_file_transfer(
+                            f, filetransfer, overall_progress_update)
+
+                        if self.is_cancelled():
+                            break
+
+                        if increment == 2:
+                            file_checksums_to_annotate.append(f.id)
+                        else:
+                            number_of_skipped_files += increment
+                except Exception as e:
+                    exception = e
+                    break
```

Comment:
Hi @lyw07 !

Hmm.. this should be logged, since more than one exception can occur, and then `exception` will be overwritten and only the last exception is re-raised.

The rule of thumb is: Never catch all exceptions without putting at least: 1) a log message and/or 2) an inline comment.

Code is moved and re-organized, therefore these blocks are risky.

Reply:
Hi @benjaoming! Thank you for the advice! I'm not sure if I understand your comment correctly, but I log the first exception here: https://github.com/lyw07/kolibri/blob/b63504fe9d8894704422fae83fce8d385c7589a9/kolibri/content/management/commands/importcontent.py#L215

Also, I couldn't find a good exception to catch, so I make the exceptions by myself.
In `transfer.py`, I raise an `ExistingTransferInProgress`:
```
def next(self):
        try:
            -> raise ExistingTransferInProgress('Transfer exception!')
            return super(FileDownload, self).next()
        except ConnectionError as e:
            logging.error("Error reading download stream: {}".format(e))
            raise
```

In `import_export_content.py`, I raise an `ImportExportException`:
```
skip_404 = kwargs.pop('skip_404')

    if (
            isinstance(e, ConnectionError) or
            isinstance(e, Timeout) or
            (isinstance(e, HTTPError) and e.response.status_code in RETRY_STATUS_CODE) or
            (isinstance(e, SSLERROR) and 'decryption failed or bad record mac' in str(e))):
        return True

    elif (
            skip_404 and
            (
                (isinstance(e, HTTPError) and e.response.status_code == 404) or
                (isinstance(e, OSError) and e.errno == 2))):
        return False

    else:
        -> raise ImportExportException(e)
```
In python 2, the logging messages are:
```
ERROR    An error occured during content import: Transfer exception!                                                                                                                    | 0/4204590 [00:00<?, ?it/s]
ERROR:kolibri.content.management.commands.importcontent:An error occured during content import: Transfer exception!                                                                        | 0/7220 [00:00<?, ?it/s]
Transfer exception!

** setting availability logs. removed for simplicity ***

WARNING  Job e1365db0f40f44b8b28df723e40e1db5 raised an exception: Traceback (most recent call last):
  File "/Users/lingyiwang/Documents/LearningEquality/kolibri/kolibri/dist/iceqube/worker/backends/inmem.py", line 75, in handle_finished_future
    result = future.result()
  File "/Users/lingyiwang/.venvs/kolibri0.10/lib/python2.7/site-packages/concurrent/futures/_base.py", line 422, in result
    return self.__get_result()
  File "/Users/lingyiwang/.venvs/kolibri0.10/lib/python2.7/site-packages/concurrent/futures/thread.py", line 62, in run
    result = self.fn(*self.args, **self.kwargs)
  File "/Users/lingyiwang/Documents/LearningEquality/kolibri/kolibri/dist/iceqube/worker/backends/inmem.py", line 149, in wrap
    raise e
ImportExportException: Transfer exception!

WARNING:iceqube.worker.backends.base:Job e1365db0f40f44b8b28df723e40e1db5 raised an exception: Traceback (most recent call last):
  File "/Users/lingyiwang/Documents/LearningEquality/kolibri/kolibri/dist/iceqube/worker/backends/inmem.py", line 75, in handle_finished_future
    result = future.result()
  File "/Users/lingyiwang/.venvs/kolibri0.10/lib/python2.7/site-packages/concurrent/futures/_base.py", line 422, in result
    return self.__get_result()
  File "/Users/lingyiwang/.venvs/kolibri0.10/lib/python2.7/site-packages/concurrent/futures/thread.py", line 62, in run
    result = self.fn(*self.args, **self.kwargs)
  File "/Users/lingyiwang/Documents/LearningEquality/kolibri/kolibri/dist/iceqube/worker/backends/inmem.py", line 149, in wrap
    raise e
ImportExportException: Transfer exception!
```

In python 3, the logging messages are:
```
ERROR    An error occured during content import: Transfer exception!     
                                                                                                               
** setting availability logs. removed for simplicity ***

WARNING  Job 7c1cdbefe84741d6aa17f4ab7396bef3 raised an exception: Traceback (most recent call last):
  File "/Users/lingyiwang/Documents/LearningEquality/kolibri/kolibri/content/management/commands/importcontent.py", line 205, in _start_file_transfer
    for chunk in filetransfer:
  File "/Users/lingyiwang/Documents/LearningEquality/kolibri/kolibri/content/utils/transfer.py", line 74, in __next__
    return self.next()
  File "/Users/lingyiwang/Documents/LearningEquality/kolibri/kolibri/content/utils/transfer.py", line 163, in next
    raise ExistingTransferInProgress('Transfer exception!')
kolibri.content.utils.transfer.ExistingTransferInProgress: Transfer exception!

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Users/lingyiwang/Documents/LearningEquality/kolibri/kolibri/dist/iceqube/worker/backends/inmem.py", line 75, in handle_finished_future
    result = future.result()
  File "/usr/local/Cellar/python/3.6.5/Frameworks/Python.framework/Versions/3.6/lib/python3.6/concurrent/futures/_base.py", line 425, in result
    return self.__get_result()
  File "/usr/local/Cellar/python/3.6.5/Frameworks/Python.framework/Versions/3.6/lib/python3.6/concurrent/futures/_base.py", line 384, in __get_result
    raise self._exception
  File "/usr/local/Cellar/python/3.6.5/Frameworks/Python.framework/Versions/3.6/lib/python3.6/concurrent/futures/thread.py", line 56, in run
    result = self.fn(*self.args, **self.kwargs)
  File "/Users/lingyiwang/Documents/LearningEquality/kolibri/kolibri/dist/iceqube/worker/backends/inmem.py", line 149, in wrap
    raise e
  File "/Users/lingyiwang/Documents/LearningEquality/kolibri/kolibri/dist/iceqube/worker/backends/inmem.py", line 145, in wrap
    return f(*args, **kwargs)
  File "/Users/lingyiwang/Documents/LearningEquality/kolibri/kolibri/dist/iceqube/common/classes.py", line 119, in y
    return func(*args, **kwargs)
  File "/Users/lingyiwang/Documents/LearningEquality/kolibri/kolibri/dist/django/core/management/__init__.py", line 131, in call_command
    return command.execute(*args, **defaults)
  File "/Users/lingyiwang/Documents/LearningEquality/kolibri/kolibri/dist/django/core/management/base.py", line 330, in execute
    output = self.handle(*args, **options)
  File "/Users/lingyiwang/Documents/LearningEquality/kolibri/kolibri/tasks/management/commands/base.py", line 99, in handle
    return self.handle_async(*args, **options)
  File "/Users/lingyiwang/Documents/LearningEquality/kolibri/kolibri/content/management/commands/importcontent.py", line 233, in handle_async
    renderable_only=options["renderable_only"])
  File "/Users/lingyiwang/Documents/LearningEquality/kolibri/kolibri/content/management/commands/importcontent.py", line 114, in download_content
    self._transfer(DOWNLOAD_METHOD, channel_id, node_ids=node_ids, exclude_node_ids=exclude_node_ids, baseurl=baseurl, renderable_only=renderable_only)
  File "/Users/lingyiwang/Documents/LearningEquality/kolibri/kolibri/content/management/commands/importcontent.py", line 180, in _transfer
    raise exception
  File "/Users/lingyiwang/Documents/LearningEquality/kolibri/kolibri/content/management/commands/importcontent.py", line 159, in _transfer
    f, filetransfer, overall_progress_update)
  File "/Users/lingyiwang/Documents/LearningEquality/kolibri/kolibri/content/management/commands/importcontent.py", line 216, in _start_file_transfer
    retry = import_export_content.retry_import(e, skip_404=True)
  File "/Users/lingyiwang/Documents/LearningEquality/kolibri/kolibri/content/utils/import_export_content.py", line 114, in retry_import
    raise ImportExportException(e)
kolibri.content.utils.import_export_content.ImportExportException: Transfer exception!

WARNING:iceqube.worker.backends.base:Job 7c1cdbefe84741d6aa17f4ab7396bef3 raised an exception: Traceback (most recent call last):
  File "/Users/lingyiwang/Documents/LearningEquality/kolibri/kolibri/content/management/commands/importcontent.py", line 205, in _start_file_transfer
    for chunk in filetransfer:
  File "/Users/lingyiwang/Documents/LearningEquality/kolibri/kolibri/content/utils/transfer.py", line 74, in __next__
    return self.next()
  File "/Users/lingyiwang/Documents/LearningEquality/kolibri/kolibri/content/utils/transfer.py", line 163, in next
    raise ExistingTransferInProgress('Transfer exception!')
kolibri.content.utils.transfer.ExistingTransferInProgress: Transfer exception!

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Users/lingyiwang/Documents/LearningEquality/kolibri/kolibri/dist/iceqube/worker/backends/inmem.py", line 75, in handle_finished_future
    result = future.result()
  File "/usr/local/Cellar/python/3.6.5/Frameworks/Python.framework/Versions/3.6/lib/python3.6/concurrent/futures/_base.py", line 425, in result
    return self.__get_result()
  File "/usr/local/Cellar/python/3.6.5/Frameworks/Python.framework/Versions/3.6/lib/python3.6/concurrent/futures/_base.py", line 384, in __get_result
    raise self._exception
  File "/usr/local/Cellar/python/3.6.5/Frameworks/Python.framework/Versions/3.6/lib/python3.6/concurrent/futures/thread.py", line 56, in run
    result = self.fn(*self.args, **self.kwargs)
  File "/Users/lingyiwang/Documents/LearningEquality/kolibri/kolibri/dist/iceqube/worker/backends/inmem.py", line 149, in wrap
    raise e
  File "/Users/lingyiwang/Documents/LearningEquality/kolibri/kolibri/dist/iceqube/worker/backends/inmem.py", line 145, in wrap
    return f(*args, **kwargs)
  File "/Users/lingyiwang/Documents/LearningEquality/kolibri/kolibri/dist/iceqube/common/classes.py", line 119, in y
    return func(*args, **kwargs)
  File "/Users/lingyiwang/Documents/LearningEquality/kolibri/kolibri/dist/django/core/management/__init__.py", line 131, in call_command
    return command.execute(*args, **defaults)
  File "/Users/lingyiwang/Documents/LearningEquality/kolibri/kolibri/dist/django/core/management/base.py", line 330, in execute
    output = self.handle(*args, **options)
  File "/Users/lingyiwang/Documents/LearningEquality/kolibri/kolibri/tasks/management/commands/base.py", line 99, in handle
    return self.handle_async(*args, **options)
  File "/Users/lingyiwang/Documents/LearningEquality/kolibri/kolibri/content/management/commands/importcontent.py", line 233, in handle_async
    renderable_only=options["renderable_only"])
  File "/Users/lingyiwang/Documents/LearningEquality/kolibri/kolibri/content/management/commands/importcontent.py", line 114, in download_content
    self._transfer(DOWNLOAD_METHOD, channel_id, node_ids=node_ids, exclude_node_ids=exclude_node_ids, baseurl=baseurl, renderable_only=renderable_only)
  File "/Users/lingyiwang/Documents/LearningEquality/kolibri/kolibri/content/management/commands/importcontent.py", line 180, in _transfer
    raise exception
  File "/Users/lingyiwang/Documents/LearningEquality/kolibri/kolibri/content/management/commands/importcontent.py", line 159, in _transfer
    f, filetransfer, overall_progress_update)
  File "/Users/lingyiwang/Documents/LearningEquality/kolibri/kolibri/content/management/commands/importcontent.py", line 216, in _start_file_transfer
    retry = import_export_content.retry_import(e, skip_404=True)
  File "/Users/lingyiwang/Documents/LearningEquality/kolibri/kolibri/content/utils/import_export_content.py", line 114, in retry_import
    raise ImportExportException(e)
kolibri.content.utils.import_export_content.ImportExportException: Transfer exception!
```

So I feel that the case that multiple exceptions are raised is handled and logged, but I could misunderstand your comment. Please let me know if I'm following in the right way. Thank you!

Reply:
Hi @lyw07 - exc3eptions are... _exceptional_ :smile: 

The main risk of catching exception without logging or any comments is that future debugging (done by our future selves or unknowing colleagues) are at risk of spending a lot of time understanding unpredicted behavior.

The rule of thumb arises from exactly these lines of code:

```
except Exception as e:
    exception = e
    break
```

Which is equivalent to:

```
except:
    break
```

You could log with `logging.debug` to avoid noisy user logs.